pkg/erofs/erofs.go

```diff
@@ -0,0 +1,729 @@
+	}
+
+	return nil
+}
+
+// FD returns the host FD of underlying image file.
+func (i *Image) FD() int {
+	return int(i.src.Fd())
+}
+
+// BytesAt returns the bytes at [off, off+n) of the image.
+func (i *Image) BytesAt(off, n uint64) ([]byte, error) {
+	size := uint64(len(i.bytes))
+	end := off + n
+	if off >= size || off > end || end > size {
+		log.Warningf("Invalid range (off: 0x%x, n: 0x%x) for image (size: 0x%x)", off, n, size)
+		return nil, linuxerr.EFAULT
+	}
+	return i.bytes[off:end], nil
+}
+
+// UnmarshalAt deserializes data from the bytes at [off, off+n) of the image.
+func (i *Image) UnmarshalAt(data marshal.Marshallable, off uint64) error {
+	bytes, err := i.BytesAt(off, uint64(data.SizeBytes()))
+	if err != nil {
+		log.Warningf("Failed to deserialize %T from 0x%x.", data, off)
+		return err
+	}
+	data.UnmarshalUnsafe(bytes)
+	return nil
+}
+
+// Inode returns the inode identified by nid.
+//
+// TODO: Ideally, we should avoid escaping objects to heap when constructing
+// objects from the image.
+func (i *Image) Inode(nid uint64) (Inode, error) {
+	inode := Inode{
+		image: i,
+		nid:   nid,
+	}
+
+	off := i.sb.NidToOffset(nid)
+	if err := i.UnmarshalAt(&inode.format, off); err != nil {
+		return Inode{}, err
+	}
+
+	var (
+		rawBlockAddr uint32
+		inodeSize    int
+	)
+
+	switch layout := inode.Layout(); layout {
+	case InodeLayoutCompact:
+		var ino InodeCompact
+		if err := i.UnmarshalAt(&ino, off); err != nil {
+			return Inode{}, err
+		}
+
+		if ino.XattrCount != 0 {
+			log.Warningf("Unsupported xattr of inode (nid=%v)", nid)
+			return Inode{}, linuxerr.ENOTSUP
+		}
+
+		rawBlockAddr = ino.RawBlockAddr
+		inodeSize = ino.SizeBytes()
+
+		inode.size = uint64(ino.Size)
+		inode.nlink = uint32(ino.Nlink)
+		inode.mode = ino.Mode
+		inode.uid = uint32(ino.UID)
+		inode.gid = uint32(ino.GID)
+		inode.mtime = i.sb.BuildTime
+		inode.mtimeNsec = i.sb.BuildTimeNsec
+
+	case InodeLayoutExtended:
+		var ino InodeExtended
+		if err := i.UnmarshalAt(&ino, off); err != nil {
+			return Inode{}, err
+		}
+
+		if ino.XattrCount != 0 {
+			log.Warningf("Unsupported xattr of inode (nid=%v)", nid)
+			return Inode{}, linuxerr.ENOTSUP
+		}
+
+		rawBlockAddr = ino.RawBlockAddr
+		inodeSize = ino.SizeBytes()
+
+		inode.size = ino.Size
+		inode.nlink = ino.Nlink
+		inode.mode = ino.Mode
+		inode.uid = ino.UID
+		inode.gid = ino.GID
+		inode.mtime = ino.Mtime
+		inode.mtimeNsec = ino.MtimeNsec
+
+	default:
+		log.Warningf("Unsupported layout 0x%x of inode (nid=%v)", layout, nid)
+		return Inode{}, linuxerr.ENOTSUP
+	}
+
+	switch dataLayout := inode.DataLayout(); dataLayout {
+	case InodeDataLayoutFlatInline:
+		// Check that whether the file data in the last block fits into
+		// the remaining room of the metadata block.
+		blockSize := i.BlockSize()
+		tailSize := uint32(inode.size) & (blockSize - 1)
+		if tailSize != 0 && tailSize < blockSize-uint32(inodeSize) {
```

Comment:
> Is the inline data stored when tailSize == blockSize-uint32(inodeSize)? Does EROFS specification requires tail data to be less than remaining metadata room?

Yeah, I think so. In erofs-utils, it will return `ENOSPC` only when inline data is bigger than remaining metadata room:

https://github.com/erofs/erofs-utils/blob/ce36273833096f1b7e828309d9b1caa37132092d/lib/cache.c#L140-L143
```c
static int erofs_bfind_for_attach(int type, erofs_off_t size,
				  unsigned int required_ext,
				  unsigned int inline_ext,
				  unsigned int alignsize,
				  struct erofs_buffer_block **bbp)
{
	const unsigned int blksiz = erofs_blksiz(&sbi);
	......

	used0 = ((size + required_ext) & (blksiz - 1)) + inline_ext;
	/* inline data should be in the same fs block */
	if (used0 > blksiz)
		return -ENOSPC;
```

This implementation doesn't support xattrs, etc yet, so `(size + required_ext)` should always be inode size.

> When data layout is InodeDataLayoutFlatInline and tail data does not fit in remaining room, should we error? Or is data layout == InodeDataLayoutFlatInline is just an indication that inline data can be present, rather than inline data is present.

Good point! You are right! I had a misunderstanding on `InodeDataLayoutFlatInline` data layout.. Thanks!